File: Data Structures & Algorithms/count-subsequences/submission-6.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int n{static_cast<int>(s.size())}, m{static_cast<int>(t.size())};
        
        vector<int> dp(m + 1);
        dp[0] = 1;
        for (int i{1}; i <= n; ++i) {
            vector<int> newDp(m + 1);
            newDp[0] = 1;
            for (int j{1}; j <= m; ++j) {
                newDp[j] = dp[j];
                if (s[i - 1] == t[j - 1]) newDp[j] += dp[j - 1];
            }
            dp = newDp;
        }
        // {
        //     i == n && j == m return 1
        //     if i > n || j > m return 0
        //     return max(self(i + 1, j), self(i, j + 1))

        // }
        return dp[m];
    }
};

File: Data Structures & Algorithms/word-break/submission-6.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        boolean[] dp = new boolean[s.length() + 1];
        int maxLen = 0;
        Set<String> set = new HashSet<>();
        for (String word : wordDict) {
            set.add(word);
            maxLen = Math.max(maxLen, word.length());
        }
        dp[s.length()] = true;
        int n = s.length();
        for (int i = n - 1; i >= 0; i--) {
            for (int r = i + 1; r <= i + maxLen && r <= n; ++r) {
                System.out.println(s.substring(i, r));
                if (set.contains(s.substring(i, r))) {
                    dp[i] = dp[i] || dp[r];
                }
            }
        }
        return dp[0];
    }
}
